presto-proxy/src/main/java/com/facebook/presto/proxy/ProxyResource.java

```diff
@@ -0,0 +1,341 @@
+            UriInfo uriInfo,
+            AsyncResponse asyncResponse,
+            Request.Builder request)
+    {
+        setupBearerToken(servletRequest, request);
+
+        for (String name : list(servletRequest.getHeaderNames())) {
+            if (isPrestoHeader(name) || name.equalsIgnoreCase(COOKIE)) {
+                for (String value : list(servletRequest.getHeaders(name))) {
+                    request.addHeader(name, value);
+                }
+            }
+        }
+
+        ListenableFuture<Response> future = executeHttp(request.build())
+                .transform(response -> buildResponse(uriInfo, response), executor)
+                .catching(ProxyException.class, ProxyResource::handleProxyException, directExecutor());
+
+        setupAsyncResponse(asyncResponse, future);
+    }
+
+    private Response buildResponse(UriInfo uriInfo, ProxyResponse response)
+    {
+        byte[] body = rewriteNextUri(response.getBody(), uri -> rewriteUri(uriInfo, uri));
+        ResponseBuilder builder = Response.ok(body);
+
+        response.getHeaders().asMap().forEach((headerName, value) -> {
+            String name = headerName.toString();
+            if (isPrestoHeader(name) || name.equalsIgnoreCase(SET_COOKIE)) {
+                builder.header(name, value);
+            }
+        });
+
+        return builder.build();
+    }
+
+    private String rewriteUri(UriInfo uriInfo, String uri)
+    {
+        return uriInfo.getAbsolutePathBuilder()
+                .replacePath("/v1/statement/next")
+                .queryParam("uri", uri)
+                .queryParam("hmac", hmac.hashString(uri, UTF_8))
+                .build()
+                .toString();
+    }
+
+    private void setupAsyncResponse(AsyncResponse asyncResponse, ListenableFuture<Response> future)
+    {
+        bindAsyncResponse(asyncResponse, future, executor)
+                .withTimeout(ASYNC_TIMEOUT, () -> Response
+                        .status(Status.BAD_GATEWAY)
+                        .type(TEXT_PLAIN_TYPE)
+                        .entity("Request to remote Presto server timed out")
+                        .build());
+    }
+
+    private FluentFuture<ProxyResponse> executeHttp(Request request)
+    {
+        return FluentFuture.from(httpClient.executeAsync(request, new ProxyResponseHandler()));
+    }
+
+    private static <T> T handleProxyException(ProxyException exception)
+    {
+        log.warn(exception, "Proxy request failed");
+        throw badRequest(Status.BAD_GATEWAY, exception.getMessage());
+    }
+
+    private void setupBearerToken(HttpServletRequest servletRequest, Request.Builder request)
+    {
+        if (!jwtKey.isPresent()) {
+            return;
+        }
+
+        X509Certificate[] certs = (X509Certificate[]) servletRequest.getAttribute(X509_ATTRIBUTE);
+        if ((certs == null) || (certs.length == 0)) {
+            throw badRequest(Status.FORBIDDEN, "No TLS certificate present for request");
```

Comment:
For password authentication, we'd need to forward it along in the proxy, which seems reasonable, but we don't do that now. It's expected for now that if you enable JWT, the proxy will require a client certificate. If we skip it, then the request to the backend server will fail because HTTPS requires authentication (if enabled).